pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -190,7 +259,17 @@ public BrokerResponseNative handleRequest(JsonNode request, @Nullable RequesterI
     if (sql == null) {
       throw new BadQueryRequestException("Failed to find 'sql' in the request: " + request);
     }
-    return handleRequest(requestId, sql.asText(), request, requesterIdentity, requestContext);
+    if (!_enableQueryCancellation) {
+      return handleRequest(requestId, sql.asText(), request, requesterIdentity, requestContext);
+    }
+    _queriesById.put(requestId, new QueryRoutingTable(sql.asText()));
```

Comment:
(minor) We might want to pass in `QueryRoutingTable` in `handleRequest()` to avoid extra lookup. It can also avoid NPE if somehow the request is early removed (probably not possible right now)

Reply:
moved inside the handleRequest() just before sending the queries out to servers, to avoid the potential failures that could happen before sending the query out to servers.